src/std/misc/list.ss

```diff
@@ -302,6 +304,36 @@
     ([a] [[a]])
     (_ (helper))))
 
+;; group consecutive clusters of n elements of the list into a list-of-lists
+;; The last element of the list returned may have fewer than n elements.
+;; : Nat (List X) -> (List (List X))
+(def (group-by n list)
+  (cond
+   ((null? list) [])
+   ((length<=n? list n) [list])
+   (else (let-values (((head tail) (split-at list n))) (cons head (group-by n tail))))))
+
+;; Given a list l of X, a key function f from X to Y, and a presumably empty table of Y to a list of X
+;; (by default an empty hash-table), add the elements in l to t, then return the list for each added
+;; key y in t of the list of elements xs with that same key y (for the equality predicate of t).
+;; Otherwise preserve the order of appearance of keys and elements for each key.
+;; : (List X) (Fun X -> Y) ?(Table Y -> (List X)) -> (List (List X))
+(def (grouping l f (t (make-hash-table)))
+  (def ys (with-list-builder (c)
+            (for-each! l
+              (lambda (x)
+                (def y (f x))
+                (def p (hash-get t y))
+                (if p
+                  (hash-put! t y (cons x p))
+                  (begin
+                    (hash-put! t y (list x))
+                    (c y)))))))
+  (map (lambda (y) (reverse (hash-get t y))) ys))
```

Comment:
does it make sense to use `reverse!` here? Why is the hash table an input parameter?

Reply:
And more importantly, what's the ownership rules for it?

Reply:
We just created `buf` above, it never escapes the function. For `reverse!` to be invalid would require some weird continuation capture and replay inside the `stop` function. But if you want to strengthen against such case, then yes it should be `reverse`.

Reply:
Lets use `reverse!` then.